keps/sig-cli/kustomize-reorder-plugins.md

```diff
@@ -0,0 +1,217 @@
+kustomize build <xxx> --reorder=kubectldelete | kubect delete -f -
+```
+
+or 
+
+```bash
+kubectl apply -k <xxx>
+kubectl delete -k <xxx>
+```
+
+### kubectl apply example:
+
+The `--reorder=kubectctlapply` option indicates to kustomize to load the `kubectlapplyordertransformer.yaml transformer config` ,
+itself loading a kustomize external plugin in charge of sorting and filtering kustomize build output.
+
+An example of configuration of such a kubectapplyordertransformer.yaml:
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: KubectlApplyOrderTransformer
+metadata:
+  name: kubectlapplyordertransformer
+kindorder:
+- Namespace
+- CustomResourceDefinition
+- ServiceAccount
+- ClusterRole
+- RoleBinding
+- ClusterRoleBinding
+- ConfigMap
+- Service
+- Deployment
+- CronJob
+- ValidatingWebhookConfiguration
+- APIService
+- Job
+- Certificate
+- ClusterIssuer
+- Issuer
+```
+
+A new transformer could also be implemented to filter resources during `kubectl apply` for instance:
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: KubectlApplyFilterTransformer
+metadata:
+  name: kubectlapplyfiltertransformer
+kindorder:
+- Namespace
+- CustomResourceDefinition
+- ServiceAccount
+- ClusterRole
+- RoleBinding
+- ClusterRoleBinding
+filter:
+- Namespace
+- CustomResourceDefinition
+```
+
+Another example could be the one based on a previous KEP:
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: KubectlApplyAnnotationBasedSorter
+metadata:
+  name: kubectlapplyannotationbasedsorter
+sortOrder:
+- matchAnnotations:
+    some-annotation-name-1: some-annotation-value-1
+    some-annotation-name-a: some-label-value-a
+- matchAnnotations:
+    some-annotation-name-1: some-annotation-value-1
+- matchAnnotations:
+    some-annotation-name-2: some-annotation-value-2
+```
+
+
+### kubectl delete example:
+
+The `--reorder=kubectctldelete` option indicates to kustomize to load a `kubectlapplyordertransformer.yaml transformer config`,
+itself loading a kustomize external plugin in charge of sorting and filtering kustomize build output.
+
+An example of configuration of such a transformer.
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: KubectlDeleteOrderTransformer
+metadata:
+  name: kubectldeleteordertransformer
+kindorder:
+- Issuer
+- ClusterIssuer
+- Certificate
+- Job
+- APIService
+- ValidatingWebhookConfiguration
+- CronJob
+- Deployment
+- Service
+- ConfigMap
+- ClusterRoleBinding
+- RoleBinding
+- ClusterRole
+- ServiceAccount
+- CustomResourceDefinition
+- Namespace
+```
+
+### Risks and Mitigations
+
+Risk: Current external plugins require the presence of the module in the XDG_CONFIG_HOME.
+Mitigation: Good documentation and recommendations about how to keep things simple.
+
+## Graduation Criteria
+
+Use customer feedback to determine if we should support:
+
+- Let the user specify the transformers configuration file in the "top" kustomization.yaml.
```

Comment:
Assume there are base and overly. Both the base and overlay contain its own order transformers. Do the resources from the base follow the order customized in base? Does the overlay order overwrite all the base order?

Reply:
done